Add locale switcher to header

tnightingale opened the pull request:
Adds a button to the header which also displays the current locale. When clicked triggers sidebar drawer to slide out with locale switcher inside.

Some possible outstanding tasks (need to decide whether these are within scope of MVP and, if so,  are within scope of this issue):

- [ ] Should `localeSwitcherOpen` reducer/store state & `openLocaleSwitcher()` actions be bundled in with rest of locale actions/state or be put alongside other general application UI state?
- [ ] Add "Need help" button to header (seems unrelated & not sure if it made it into MVP?)
- [ ] Keep drawer open through application restart (do we want this? I suspect it's more important to persist user's current page)
- [ ] Short-hand and/or localized names for locales (eg: 'EN' instead of 'English')
- [x] Make sure header button icons correctly mirror in RTL
- [ ] ... what else?

bensinjin commented:
This looks good. I think we probably want to group our UI related state together but that's just my opinion. I think we don't want to persist the drawer being open (page is important I agree) although for it to feel consistent would be nice. I would almost lean towards the LTR language switching to work like the RTL switch meaning the drawer closes when you select a language. I'm not sure how hard that is or if it goes too far against how people expect the drawer to work (I personally think closing is a sorta unintuitive).  Finally I'm not sure if this is meant to mirror the designs but it could def. use some love in Android.

tnightingale commented:
> I think we probably want to group our UI related state together but that's just my initial thought.

Yes, I think so too but I get stuck trying to define what constitutes UI state and what is not. I can see a scenario where we create a `userInterfaceStore` for all things UI that it just becomes a 
dumping ground for a bunch of unrelated state. :-\ In the meantime, I think our current state grouping/organization is sufficient.

> I would almost lean towards the LTR language switching to work like the RTL switch meaning the drawer closes when you select a language.

My concern was that it might be disorientating to have the drawer disappear. But I actually think that this is the best approach.

> Finally I'm not sure if this is meant to mirror the designs but it could def. use some love on Android.

I assume you're referring to the fact that it sits under the OS's status/notification bar. This is currently a problem application-wide (it's even worse with an iPhone X's 'notch'). It needs to be addressed with a top-level style.

rasmus-storjohann-PG commented:
Did you notice there is a story #106? All work to get to the MVP should be covered in stories under #104, unless there's something we've overlooked. Please check #104 carefully before writing new stories.

tnightingale commented:
No I did not, this is in response to #100 (which I did not create, it was assigned to me during the last sprint planning meeting).

bensinjin commented:
> I assume you're referring to the fact that it sits under the OS's status/notification bar. This is currently a problem application-wide (it's even worse with an iPhone X's 'notch'). It needs to be addressed with a top-level style.

I just tried it on Android the and the buttons are square and blue and it generally does not match designs provided.

rasmus-storjohann-PG commented:
"Need help" button is out of scope as per https://docs.google.com/document/d/1130BnFQS0T4SrE7lA12bZjjTZuY8Fr7xzObunFxNkJE/edit#

rasmus-storjohann-PG commented:
@tnightingale totally my bad, I thought I was looking at a story, not a PR.

tnightingale commented:
Good points @bensinjin I will update styling to match mocks (#118).

rasmus-storjohann-PG commented:
`localeSwitcherOpen` is not part of the locale store, it's either part of some localeSwitcher store or just stored in the component. I'm interested in exploring when component state should just live in the component. To be clear, I'm not a redux-store-for-everything-no-class-based-components zealot.

rasmus-storjohann-PG commented:
I would not expect the drawer to remain open after app restart ... I don't think. Do whatever is easier for now.

rasmus-storjohann-PG commented:
Localized names (English) for locales in the menu, definitely.

tnightingale commented:
> I would not expect the drawer to remain open after app restart ... I don't think. Do whatever is easier for now.

But should it remain open after a same-layout language is selected?

It seems like the consensus is trending towards "no" which means that we can move the `localeSwitcherOpen` state into component state.

rasmus-storjohann-PG commented:
Yeah, I think the drawer closes when you select a language, so we don't need an additional click to confirm the selection..

tnightingale commented:
> Localized names (English) for locales in the menu, definitely.

So, to clarify: locale names should be in English, eg:
- English
- Arabic
- Chinese

Not:
- English
- عربى
- 中文

rasmus-storjohann-PG commented:
For the store organization, really pay attention to the structure that makes the components easy to test, that will typically be the best design. I anticipate that this will likely move us more towards a `fooComponentStore` and `barComponentStore` as opposed to `allUIStore.foo` and `allUIStore.bar`.

rasmus-storjohann-PG commented:
Locale names should be in the current locale, so "English" or "Anglais" depending,

rasmus-storjohann-PG commented:
Hang on a second, I spoke too soon. We should not go against the mocks so easily, I'm guessing Cherry put thought into this. And this is what the mockis have looked like all along. I think we should proceed according to the mocks.

rasmus-storjohann-PG commented:
Yup, confirmed with Cherry, follow the mocks, my bad.

rasmus-storjohann-PG commented on src/stores/__tests__/locale.test.ts:
```diff
@@ -106,6 +114,16 @@ describe('the reducer', () => {
+        const theStore = locale.reducer();
+        const theAction = {
+            type: constants.OPEN_LOCALE_SWITCHER as typeof constants.OPEN_LOCALE_SWITCHER,
+            payload: { open: true },
```
`payload: { open: aBoolean() },` then you test both paths in one test

rasmus-storjohann-PG commented:
Actually, I don't think you're done. Will not merge yet. Bug me if anything is unclear around this.

tnightingale commented:
Ok, I have:
 - moved the drawer logic into component state
 - tidied up styling using proper React Native and Native Base components (FYI Native Base's `<List />` is deprecated in favour of RN's `<FlatList />` and `<SectionList />` components)
 - updated the locale labels to be their localized (per-language) versions

I have implemented this with RN's `<SectionList />` as it's supposed to give us "sticky-headers" but it doesn't seem to work. I'm interested to hear if anyone knows what's up with that?
There's currently a bunch of dummy locales registered (A, B, C, D, ..., etc) to demonstrate the list scrolling (and supposed to show "sticky-headers").

Please review :)

rasmus-storjohann-PG commented:
Change the commit messages to reference #106 rather than #100, sorry for the confusion on that.

rasmus-storjohann-PG commented:
I like the UX of the drawer closing immediately on selecting a language, it's consistent whether the app switches directionality or not, which is good.

rasmus-storjohann-PG commented:
There's an arrow on each entry in the select language menu which is not there in the mocks, is that a side effect of the SecitonList?

rasmus-storjohann-PG commented on src/application/store.ts:
```diff
@@ -16,8 +16,26 @@ import zhMessages from '../../locale/zh/messages';
+    { code: 'o', label: 'O', catalog: zhMessages, isRTL: false },
+    { code: 'p', label: 'P', catalog: enMessages, isRTL: false },
+    { code: 'q', label: 'Q', catalog: arMessages, isRTL: true },
+    { code: 'r', label: 'R', catalog: zhMessages, isRTL: false },
```
I'm not sure why these extra entries are needed. UX needs to look OK with a small set of supported languages, the MVP will have at most three.

tnightingale replied:
To demonstrate the acceptance criteria in #106:

> List of languages, with scroll behaviour, note how Arabic slides behind the "English" label:

I mentioned this in my comment above:

> There's currently a bunch of dummy locales registered (A, B, C, D, ..., etc) to demonstrate the list scrolling (and supposed to show "sticky-headers").

The sticky-headers functionality, if it worked as advertised, would provide the "slides behind" effect described in AC.

> Also, we need to make sure that the UX works with a smaller number of supported locale.

It works fine when there's only 3 languages, the list just does not scroll. I will remove the extra languages.

rasmus-storjohann-PG replied:
OK, now I understand, thanks

rasmus-storjohann-PG commented on src/components/language_switcher/connected_language_switcher.ts:
```diff
@@ -4,12 +4,14 @@ import { LanguageSwitcher, Props, Actions } from './language_switcher';
+    const locale = selectLocale(store);
+    const currentLocale = locales.find((aLocale: LocaleInfo) => locale.code === aLocale.code);
+    const availableLocales = locales.filter((aLocale: LocaleInfo) => locale.code !== aLocale.code);
+    return { currentLocale, availableLocales };
```
This should be in a selector, there should be no logic in `mapStateToProps()`.

tnightingale replied:
I'm not sure I agree with this policy; it is arbitrary and therefore prevents situations where it might be appropriate to have component-specific logic (though, to be clear, it should definitely be tested).

rasmus-storjohann-PG commented on src/components/language_switcher/language_switcher.tsx:
```diff
@@ -16,21 +17,56 @@ export type LanguageSwitcherProps = Props & Actions;
-        </View>
+        <ListItem key={item.code} onPress={(): void => item.setLocale(item.code)}>
+            <Left><Text>{item.label}</Text></Left>
+            <Right><Icon name={I18nManager.isRTL ? 'arrow-back' : 'arrow-forward'} /></Right>
```
OK, I understand the arrows now ...

rasmus-storjohann-PG commented on src/components/main/header.tsx:
```diff
@@ -1,24 +1,37 @@
-    </Header>;
+export interface UiActions {
+    readonly onLanguageSelect: () => void;
+}
```
Seems to me that these two actions interfaces can be combined into one.

tnightingale replied:
These need to be separate as `Header#Actions` is exported and combined with the other Main component `Actions` & `Props` (from `footer` and `mainPageSwitcher` in `src/main/main.tsx`). This union type is used to create the connected component which has no knowledge of the internal state of its subcomponents.

Perhaps we need to improve our naming conventions to help clarify this?

rasmus-storjohann-PG replied:
OK, makes sense. Yeah, we'll need to look at naming in this context... post MPV

rasmus-storjohann-PG commented on src/components/main/main.tsx:
```diff
@@ -1,16 +1,51 @@
+import { ConnectedLanguageSwitcher } from '../language_switcher/connected_language_switcher';
+
+interface DrawerState {
+    readonly isDrawerOpen: boolean;
```
So with this state in the component, you don't need the state in the store, right?

tnightingale replied:
Not sure what you're getting at, the state has been removed from the store.

tnightingale replied:
Sorry, I see now that I hadn't completely removed traces from the store. Should be gone now.

rasmus-storjohann-PG commented on src/stores/__tests__/helpers/locale_helpers.ts:
```diff
@@ -122,6 +123,7 @@ export class LocaleStoreBuilder {
             fallback: this.fallback,
             loading: this.loading,
             errorMessage: this.errorMessage,
+            localeSwitcherOpen: this.localeSwitcherOpen,
```
No longer needed, I think

tnightingale replied:
Good catch, will remove.

tnightingale commented on src/components/language_switcher/language_switcher.tsx:
```diff
@@ -16,21 +17,71 @@ export type LanguageSwitcherProps = Props & Actions;
+    );
+};
+
+type SectionHeaderInfo = {
```
I'm not particularly happy with these types but unfortunately React-Native's types ([@types/react-native](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/types/react-native/index.d.ts#L3855-L3991) don't fully match the API (and the API itself is quite convoluted). Happy to take suggestions if anyone has a better way of organizing this.

tnightingale commented:
Rebased and removed the extra locales

rasmus-storjohann-PG commented on src/components/language_switcher/language_switcher.tsx:
```diff
@@ -51,22 +48,40 @@ type SectionItemInfo = {
+    return (locale: LocaleInfo): LocaleListItem => {
+        return { ...locale, onPress: (): void => onPress(locale.code) };
+    };
+}
```
Nice, I've been looking for a way to do this

tnightingale replied:
Yeah you can actually avoid the extra overhead of creating per-item functions and just access it from the `section` object that's also passed to the itemRender method but sadly RN's types don't include that argument (but it's definitely there).
